src/cogent3/util/warning.py

```diff
@@ -68,3 +69,46 @@ def discontinued(_type, name, version, reason=None, stack_level=3):
     with catch_warnings():
         simplefilter("always")
         _warn(msg, DeprecationWarning, stacklevel=stack_level)
+
+
+def deprecate(
+    alternate: Union[Callable[..., Any], None] = None,
+    version: str = None,
+    reason: str = None,
+) -> Callable[..., Any]:
+    """
+    Decorator to mark a function as deprecated and optionally to be replaced by an
+    alternate function.
+
+    Args:
+        alternate (Callable[..., Any], optional): The alternate function to be called
+            instead of the deprecated function. Defaults to None.
+        version (str, optional): Specifies when the deprecated function will be removed.
+            Defaults to None.
+        reason (str, optional): The reason for deprecating the function. Defaults to None.
+
+    The decorator will call the alternate function when the decorated function is called.
+    If an alternate function is not provided, or is not a callable function
+    then the original function will be called instead.
+    """
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        def wrapper(*args: Any, **kwargs: Any) -> Callable[..., Any]:
+
+            deprecated(
```

Comment:
if it's to decorate functions, it should also decorate methods and that's where `functools.wraps` won't work in terms of correctly preserving the method name and other metadata (docstring, annotations).